core/src/main/java/hudson/model/Actionable.java

```diff
@@ -162,9 +180,20 @@ public Action getAction(int index) {
      * @see #getActions(Class)
      */
     public <T extends Action> T getAction(Class<T> type) {
-        for (Action a : getAllActions())
-            if (type.isInstance(a))
+        // Shortcut: if the persisted list has one, return it.
+        for (Action a : getActions()) {
+            if (type.isInstance(a)) {
                 return type.cast(a);
+            }
+        }
+        // Otherwise check transient factories.
+        for (TransientActionFactory<?> taf : TransientActionFactory.factoriesFor(getClass(), type)) {
+            for (Action a : createFor(taf)) {
```

Comment:
Also catch/suppress exceptions? Not so good for performance, but commonly we should not trust extension points.